File: typescript/src/ledger/canonical.ts
import { createHash } from 'node:crypto';

// Deterministic canonical JSON serialization for hashing (design audit-integrity §2:
// "決定的なフィールド順の canonical serialization に対してハッシュ" for verification
// reproducibility). Keys sorted recursively; no insignificant whitespace. Numbers and
// strings use JSON.stringify's canonical forms. undefined-valued keys are omitted.
export function canonicalize(value: unknown): string {
  return serialize(value);
}

function serialize(value: unknown): string {
  if (value === null) return 'null';
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) throw new Error('cannot canonicalize non-finite number');
    return JSON.stringify(value);
  }
  if (typeof value === 'boolean' || typeof value === 'string') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(serialize).join(',')}]`;
  if (typeof value === 'object') {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj)
      .filter((k) => obj[k] !== undefined)
      .sort();
    const body = keys.map((k) => `${JSON.stringify(k)}:${serialize(obj[k])}`).join(',');
    return `{${body}}`;
  }
  throw new Error(`cannot canonicalize value of type ${typeof value}`);
}

export function sha256Hex(input: string): string {
  return createHash('sha256').update(input, 'utf8').digest('hex');
}

// params_hash helper — tools hash their (masked) params, never the raw values (§4).
export function hashParams(params: unknown): string {
  return `sha256:${sha256Hex(canonicalize(params))}`;
}

File: python/src/auditable_mcp/canonical.py
"""Deterministic canonical JSON serialization and hashing.

Canonical form rules:
- Object keys sorted recursively.
- No insignificant whitespace.
- Non-ASCII characters preserved.
- Null values preserved.
"""

import json
from hashlib import sha256

# JSON separators with no insignificant whitespace (matches JSON.stringify output).
_SEPARATORS = (',', ':')


def canonicalize(value: object) -> str:
    """Serialize a JSON-compatible value to its canonical string form.

    Args:
        value: Any JSON-compatible value (dict, list, str, int, bool, None).

    Returns:
        The canonical JSON string: keys sorted recursively, no whitespace, non-ASCII kept.
    """
    return json.dumps(value, sort_keys=True, separators=_SEPARATORS, ensure_ascii=False)


def sha256_hex(data: str) -> str:
    """Return the hex-encoded SHA-256 of ``data`` encoded as UTF-8."""
    return sha256(data.encode('utf-8')).hexdigest()


def hash_params(params: object) -> str:
    """Return the `sha256:<hex>` hash of the canonical form of `params`."""
    return f'sha256:{sha256_hex(canonicalize(params))}'
